Replacement rules for Bokeh classes

mbrown1413 opened the pull request:
Use replacement rules in docs and docstrings to reference classes. Since the sphinx `rst_prolog` contains so many replacement rules now, I also separated it into its own file.

- [ ] issues: addresses #8448

mbrown1413 commented:
Fixed some trailing whitespace. All checks are passing now.

bryevdv commented:
Links to `Scatter` (e.g. in the 1.0 release notes highlights) are not working for some reason.

mbrown1413 commented on sphinx/prolog.rst:
```diff
@@ -0,0 +1,524 @@
+.. |MultiValuedDict| replace:: :py:class:`~bokeh.util.datatypes.MultiValuedDict`
+.. |Options| replace:: :py:class:`~bokeh.util.options.Options`
+.. |BokehDeprecationWarning| replace:: :py:class:`~bokeh.util.warnings.BokehDeprecationWarning`
+.. |BokehUserWarning| replace:: :py:class:`~bokeh.util.warnings.BokehUserWarning`
```
It's sorted by fully qualified class name including the package, but it might make sense to sort just by class name. Either way, I'll put some blank lines to make it easier to navigate.

mbrown1413 commented:
I actually noticed that with ``Model``. I think there is a problem with some classes being linked because they aren't included in the module's ``__all__``. I'll have more time to look into this later in the week.

bryevdv commented:
(FYI trying to push a 1.0.4 release soon so temp quit this build to clear up CI build workers)

mbrown1413 commented:
Sorry this has taken a while. I pushed a commit (and rebased) fixing some docs references. `Scatter` and `XYGlyph` were not actually documented, so I added them and now the links work. There are some other classes that I had replacement rules for that looked internal, so I removed those replacement rules. There are still some I'm not sure about:

These aren't in their module's `__all__`, so they don't appear in the reference:
  * `Model`
  * `MetaModel`
  * `Event`

Is there a reason they aren't in `__all__`? Should they be put into the documentation without modifying `__all__`?

These are all properties that aren't documented, but they aren't referenced elsewhere in the docs either:
  * `ContainerProperty`
  * `DeserializationError`
  * `PrimitiveProperty`
  * `ColumnDataPropertyDescriptor`
  * `PropertyDescriptor`
  * `PropertyValueColumnData`
  * `PropertyValueContainer`
  * `PropertyValueDict`
  * `PropertyValueList`

These are also properties that aren't documented, but they are all referenced in release notes:
  * `ParameterizedProperty`
  * `Property`
  * `PropertyDescriptorFactory`
  * `BasicPropertyDescriptor`
  * `DataSpecPropertyDescriptor`
  * `UnitsSpecPropertyDescriptor`

Should these be added to the reference documentation?

bryevdv commented:
Hrmm `Model` and `Event` should probably be in their respective `__all__`. All the other things including `MetaModel` are well outside anything users should be interacting with at all.

bryevdv commented:
I'm presently subitting a PR to add `Model` and `Event`

mbrown1413 commented:
Ok, great. When that's merged I can rebase and test that the links work correctly. Then this should be good to merge.

bryevdv commented:
PR is merged now

mbrown1413 commented:
Just rebased and everything looks good to me. `Model` links work. There actually aren't any links to `Event`, but its documentation is present now.

bryevdv commented:
@mbrown1413 things are green now, however I do notice that this *significantly* increases the docs build time (almost doubles on travisCI). I think it might be worth being a bit more judicious, because looking a bit more carefully do see a fair number of things that are not really-user facing in the rst prolog. I think it's worth making a pass over this an commenting on what can be removed, but my allotted time for Bokeh this week is almost up so it will still be a little while before we can push this to the end. Thanks for your patience!

mbrown1413 commented:
I didn't even think about this affecting the build time, so good catch.

Another option is that I could remove replacement rules that aren't used. This might remove some user facing things that just aren't referenced yet though, so if you want to comment on what should be removed, that's fine too. No rush on this now that the hard part is done on my side.

bryevdv commented:
@mbrown1413 I am going to go ahead and close this as the code drift/conflicts have grown quite substantial. I do really like the idea of increasing the cross referencing throughout the docs, it just needs more investigation to figure out any compromises to keep the build times in check. I'm sorry I was not able to look at things before these conflicts arose, I would be happy to collaborate on a new PR if you are ever inclined (at this point it should be against `landing-2.0` or after 2.0 is released)
